File: apps/server/src/academic/degrees/dtos.rs
use crate::academic::{AcademicId, DegreeKind};

use chrono::NaiveDate;
use serde::{Deserialize, Serialize};
use validator::Validate;

#[derive(Debug, Serialize, Deserialize, Validate)]
#[serde(rename_all = "camelCase")]
pub struct CreateDegreeDto {
    pub academic_id: AcademicId,

    #[validate(length(
        min = 1,
        max = 255,
        message = "El nombre debe tener entre 1 y 255 caracteres"
    ))]
    pub name: String,

    #[validate(length(
        min = 1,
        max = 255,
        message = "La universidad debe tener entre 1 y 255 caracteres"
    ))]
    pub university: String,

    pub obtained_at: NaiveDate,
    pub kind: DegreeKind,

    #[validate(length(
        min = 2,
        max = 2,
        message = "El código de país debe tener 2 caracteres"
    ))]
    pub country_code: String,
}

#[derive(Debug, Serialize, Deserialize, Validate)]
#[serde(rename_all = "camelCase")]
pub struct UpdateDegreeDto {
    #[validate(length(
        min = 1,
        max = 255,
        message = "El nombre debe tener entre 1 y 255 caracteres"
    ))]
    pub name: Option<String>,

    #[validate(length(
        min = 1,
        max = 255,
        message = "La universidad debe tener entre 1 y 255 caracteres"
    ))]
    pub university: Option<String>,
    pub obtained_at: Option<NaiveDate>,

    #[validate(length(
        min = 2,
        max = 2,
        message = "El código de país debe tener 2 caracteres"
    ))]
    pub country_code: Option<String>,
}

File: apps/client/src/lib/academic/degrees/dtos.ts
import * as v from "valibot"
import { DEGREE_KIND, type DegreeKind } from "./enums"

export interface Degree {
	id: string
	academicId: string
	name: string
	university: string
	obtainedAt: string
	kind: DegreeKind
	countryCode: string
}

export const createDegreeSchema = v.object({
	academicId: v.pipe(v.string(), v.nonEmpty("El académico es obligatorio")),
	name: v.pipe(
		v.string(),
		v.minLength(1, "El nombre debe tener entre 1 y 255 caracteres"),
		v.maxLength(255, "El nombre debe tener entre 1 y 255 caracteres"),
	),
	university: v.pipe(
		v.string(),
		v.minLength(1, "La universidad debe tener entre 1 y 255 caracteres"),
		v.maxLength(255, "La universidad debe tener entre 1 y 255 caracteres"),
	),
	obtainedAt: v.pipe(v.string(), v.nonEmpty("La fecha de obtención es obligatoria")),
	kind: v.picklist(DEGREE_KIND, "Seleccione un tipo de grado"),
	countryCode: v.pipe(v.string(), v.length(2, "El código de país debe tener 2 caracteres")),
})

export type CreateDegreeDto = v.InferInput<typeof createDegreeSchema>

export const updateDegreeSchema = v.object({
	name: v.optional(
		v.pipe(
			v.string(),
			v.minLength(1, "El nombre debe tener entre 1 y 255 caracteres"),
			v.maxLength(255, "El nombre debe tener entre 1 y 255 caracteres"),
		),
	),
	university: v.optional(
		v.pipe(
			v.string(),
			v.minLength(1, "La universidad debe tener entre 1 y 255 caracteres"),
			v.maxLength(255, "La universidad debe tener entre 1 y 255 caracteres"),
		),
	),
	obtainedAt: v.optional(v.string()),
	countryCode: v.optional(
		v.pipe(v.string(), v.length(2, "El código de país debe tener 2 caracteres")),
	),
})

export type UpdateDegreeDto = v.InferInput<typeof updateDegreeSchema>
